include/mneme/plan.hpp

```diff
@@ -2,24 +2,154 @@
 #define MNEME_PLAN_H_
 
 #include <cstddef>
+#include <memory>
+#include <numeric>
+#include <optional>
+#include <utility>
 #include <vector>
 
 #include "displacements.hpp"
+#include "span.hpp"
+#include "tagged_tuple.hpp"
 
 namespace mneme {
 
 class Plan {
 public:
-    Plan(std::size_t numElements) : dofs(numElements, 0) {}
+    using LayoutType = Displacements<std::size_t>;
+    explicit Plan(std::size_t numElements) : dofs(numElements, 0) {}
 
     void setDof(std::size_t elementNo, std::size_t dof) { dofs[elementNo] = dof; }
 
-    Displacements<std::size_t> getLayout() { return Displacements(dofs); }
+    void resize(std::size_t newSize) { dofs.resize(newSize); }
+    [[nodiscard]] LayoutType getLayout() const { return Displacements(dofs); }
 
 private:
     std::vector<std::size_t> dofs;
 };
 
+class Layer {
+public:
+    constexpr Layer() = default;
+    constexpr Layer(std::size_t numElements, std::size_t offset)
+        : numElements(numElements), offset(offset) {}
+    std::size_t numElements = 0;
+    std::size_t offset = 0;
+};
+
+class LayeredPlanBase {};
+
+template <typename... Layers> class LayeredPlan : public LayeredPlanBase {
```

Comment:
Could you add a short doxygen string which briefly describes what this class does and which includes a usage example?